ttnn/cpp/ttnn/operations/experimental/transformer/rotary_embedding_llama_fused_qk/device/rotary_embedding_llama_fused_qk_device_operation.cpp

```diff
@@ -0,0 +1,107 @@
+
+    TT_FATAL(q_input_tensor.get_logical_shape()[-1] == k_input_tensor.get_logical_shape()[-1], "Q input tensor and K input tensor must have same head dimensions");
+    uint32_t head_dim = q_input_tensor.get_logical_shape()[-1];
+    TT_FATAL(head_dim <= 128 || std::get<ttnn::WormholeComputeKernelConfig>(this->compute_kernel_config).fp32_dest_acc_en == false, "If head_dim is > 128, fp32_dest_acc_en must be False");
+
+    // Check that head_dim is a multiple of 32
+    TT_FATAL(head_dim % TILE_WIDTH == 0, "Head dim must be a multiple of TILE_WIDTH");
+
+    TT_FATAL(q_input_tensor.memory_config().memory_layout == this->q_output_mem_config.memory_layout, "Q Input tensor and Q output tensor must have same memory layout");
+    TT_FATAL(k_input_tensor.memory_config().memory_layout == this->k_output_mem_config.memory_layout, "K Input tensor and K output tensor must have same memory layout");
+
+    // check that q and k have same batch size and lesser that equal to 32
+    uint32_t q_batch_size = q_input_tensor.get_logical_shape()[1];
+    uint32_t k_batch_size = k_input_tensor.get_logical_shape()[1];
+    TT_FATAL(q_batch_size == k_batch_size, "Q and K must have the equal batch size");
+    TT_FATAL(q_batch_size <= 32, "Q and K must have batch size less than or equal to 32, due to parallelization over core-grid of 64");
+    uint32_t q_num_cores = q_input_tensor.shard_spec()->grid.bounding_box().grid_size().x * q_input_tensor.shard_spec()->grid.bounding_box().grid_size().y;
+    uint32_t k_num_cores = k_input_tensor.shard_spec()->grid.bounding_box().grid_size().x * k_input_tensor.shard_spec()->grid.bounding_box().grid_size().y;
+    TT_FATAL(q_num_cores + k_num_cores <= 64, "Q and K must not exceed max core grid size of 64");
+
+    bool is_overlap = q_input_tensor.shard_spec()->grid.intersects(k_input_tensor.shard_spec()->grid);
+    TT_FATAL(!is_overlap, "Q and K must not overlap");
+
+    // Check that cos and sin have same dims
+    TT_FATAL(cos.get_logical_shape() == sin.get_logical_shape(), "Cos and Sin dims must match");
+    uint32_t cos_sin_batch_size = cos.get_logical_shape()[1];
+    TT_FATAL(cos_sin_batch_size == (q_batch_size + k_batch_size), "Cos and Sin are repeated for Q and K, so they must have the same batch size as the sum of Q and K batch sizes");
+
+    // Checks for transformation matrix
+    uint32_t trans_mat_num_cores = trans_mat.shard_spec()->grid.bounding_box().grid_size().x * trans_mat.shard_spec()->grid.bounding_box().grid_size().y;
+    TT_FATAL(trans_mat_num_cores >= (q_num_cores + k_num_cores), "Transformation matrix is repeated for Q and K must be sharded over core grid of Q and K");
+    TT_FATAL(trans_mat.shard_spec()->shape[0] == TILE_HEIGHT, "Transformation matrix must have 3rd dim equal to TILE_HEIGHT");
+    TT_FATAL(trans_mat.shard_spec()->shape[1] == TILE_WIDTH, "Transformation matrix must have 4rd dim equal to TILE_WIDTH");
```

Comment:
Please add more information to the exception. Like shape[0] value and expected TILE.

Reply:
Updated, I hope its more clear now.